geth/jail/cell_test.go

```diff
@@ -91,43 +98,83 @@ func (s *CellTestSuite) TestCellFetchRace() {
 
 	err := cell.Set("__captureSuccess", func(res otto.Value) { dataCh <- res })
 	s.NoError(err)
+
 	err = cell.Set("__captureError", func(res otto.Value) { errCh <- res })
 	s.NoError(err)
 
-	// run JS code for fetching valid URL
-	_, err = cell.Run(`fetch('` + server.URL + `').then(function(r) {
-		return r.text()
+	fetchCode := `fetch('%s?i=%d').then(function(r) {
+ 		return r.text()
 	}).then(function(data) {
 		__captureSuccess(data)
 	}).catch(function (e) {
 		__captureError(e)
-	})`)
+	})`
+
+	for i := 0; i < requestCount; i++ {
+		_, err = cell.Run(fmt.Sprintf(fetchCode, server.URL, i))
+		s.NoError(err)
+	}
+
+	for i := 0; i < requestCount; i++ {
+		select {
+		case data := <-dataCh:
+			// Mark the request as successful.
+			expected[data.String()] = true
+		case _ = <-errCh:
+			s.Fail("fetch failed to complete the request")
+		case <-time.After(5 * time.Second):
+			s.Fail("test timed out")
+			return
+		}
+	}
+
+	// Make sure every request was completed successfully.
+	for i := 0; i < requestCount; i++ {
+		s.Equal(expected[fmt.Sprintf("%d", i)], true)
+	}
+
+	// There might be some tasks about to call `ready`,
+	// add a little delay before `TearDownTest` closes the loop.
+	time.Sleep(100 * time.Millisecond)
+}
+
+func (s *CellTestSuite) TestCellFetchErrorRace() {
+	cell := s.cell
+	dataCh := make(chan otto.Value, 1)
+	errCh := make(chan otto.Value, 1)
+
+	err := cell.Set("__captureSuccess", func(res otto.Value) { dataCh <- res })
+	s.NoError(err)
+	err = cell.Set("__captureError", func(res otto.Value) { errCh <- res })
 	s.NoError(err)
 
-	// run JS code for fetching invalid URL
-	_, err = cell.Run(`fetch('http://👽/nonexistent').then(function(r) {
+	// Find a free port in localhost
+	freeportListener, err := net.Listen("tcp", "127.0.0.1:0")
+	s.NoError(err)
```

Comment:
if you want this test to exit after this error then s.Require().NoError should be used. otherwise, freeportListener might be not initialized, right?